# Importing Libraries

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import os

# Importing Dataframes

In [2]:
# Setting Variables
path = r'C:\Users\Saikat Halder\Documents\CareerFoundry\Data Immersion\Achievement 4\Instacart Basket Analysis 01182024'
path_mac = r'/Users/saikathalder/CareerFoundry/Data Immersion/Achievement 4/Instacart Basket Analysis 01182024'
# Importing Merged Dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path_mac, '02 Data', 'Prepared Data', 'orders_products_merged_all_dataframes.pkl'))

In [4]:
# Verifying dataframe
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  
0                     3.0                  5          0   both  
1                    20.0                  1          1   both  
2                     6.0                 20          0   both  
3                     0.0                 10          0   both  
4                     9.0                 11          1   both

In [5]:
#Dropping '_merge' column
df_ords_prods_merge = df_ords_prods_merge.drop(columns = '_merge')

In [6]:
#Verying the result
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered  
0                     3.0                  5          0  
1                    20.0                  1          1  
2                     6.0                 20          0  
3                     0.0                 10          0  
4                     9.0                 11          1

# 1. Create 'price_label' and 'busiest_day' columns

## Creating 'price_label' Column and Ranking products based on their prices

### Using If-Statement and LOC() Function for the following Price Ranges

In [7]:
# Low range product where prices are below 5
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_label'] = 'Low range products'
# Mid range products where prices are between 15 and 5
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_label'] = 'Mid range products'
## High range products where price is more than 15
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_label'] = 'High range products'


In [8]:
#Frequency Count
df_ords_prods_merge['price_label'].value_counts(dropna = True)

price_label
Mid range products     21860860
Low range products     10126321
High range products      417678
Name: count, dtype: int64

In [9]:
# Verifying if the new new column is added
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered         price_label  
0                     3.0                  5          0  Mid range products  
1                    20.0                  1          1  Mid range products  
2                     6.0                 20          0  Mid range products  
3                     0.0                 10          0  Mid range products  
4                     9.0                 11          1  Mid range products

## Creating 'busiest_day' column and ranking order_day_of_week depending how busy they were

### Using For Loops

In [10]:
# Frequency Count for order_day_of_week
df_ords_prods_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [11]:
# Using for loops to create a new list

result = []

for x in df_ords_prods_merge['order_day_of_week']:
    if x == 0:
        result.append('Busiest day')
    elif x == 4:
        result.append('Least busy')
    else:
        result.append('Regulary busy')

In [12]:
# Verifying the list

result

['Regulary busy',
 'Regulary busy',
 'Busiest day',
 'Regulary busy',
 'Least busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Least busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regulary busy',
 'Regulary busy',
 'Busiest day',
 'Regulary busy',
 'Regulary busy',
 'Least busy',
 'Regulary busy',
 'Busiest day',
 'Regulary busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Regulary busy',
 'Busiest day',
 'Regulary busy',
 'Regulary 

In [13]:
## Comibining with the ords_prods_merge
df_ords_prods_merge['busiest_day'] = result
# Frequency Count
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regulary busy    22416875
Busiest day       6204182
Least busy        3783802
Name: count, dtype: int64

In [14]:
# Verifying if the column was added to the dataframe
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered         price_label  \
0                     3.0                  5          0  Mid range products   
1                    20.0                  1          1  Mid range products   
2                     6.0                 20          0  Mid range products   
3                     0.0                 10          0  Mid range products   
4                     9.0                 11          1  Mid range products   

     busiest_day  
0  Regulary busy  
1  Regulary busy  
2    Busiest day  
3  Regulary busy  
4     Least busy

# 2. Updating the 'busiest_day' to 'busiest_days' and Ranking top 2 Busiest days and Top 2 Slowest Days

## Changing the column name to 'busiest_days'

In [15]:
# Changing the column name
df_ords_prods_merge.rename(columns = {'busiest_day':'busiest_days'}, inplace = True)

In [16]:
# Verifying if the column name changed
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered         price_label  \
0                     3.0                  5          0  Mid range products   
1                    20.0                  1          1  Mid range products   
2                     6.0                 20          0  Mid range products   
3                     0.0                 10          0  Mid range products   
4                     9.0                 11          1  Mid range products   

    busiest_days  
0  Regulary busy  
1  Regulary busy  
2    Busiest day  
3  Regulary busy  
4     Least busy

## Creating a new category for busiest days, slowest days and regularly busy

In [17]:
# Frequency Count for order_day_of_week
df_ords_prods_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [18]:
# Using For Loops for creating a new list where 0 OR 1 are busiest days, 3 OR 4 are slowest dayys and else as regularly busy

result_new = []

for y in df_ords_prods_merge['order_day_of_week']:
    if y == 0 or y == 1:
        result_new.append('Busiest days')
    elif y == 4 or y == 3:
        result_new.append('Slowest days')
    else:
        result_new.append('Regularly busy')

In [19]:
# # Comibining with the ords_prods_merge
df_ords_prods_merge['busiest_days'] = result_new

# 3. Check the values for this new column for accuracy

In [40]:
# Frequency Count
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

ValueError: Grouper for 'busiest_days' not 1-dimensional

In [20]:
# Verifying column using the head() function
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered         price_label  \
0                     3.0                  5          0  Mid range products   
1                    20.0                  1          1  Mid range products   
2                     6.0                 20          0  Mid range products   
3                     0.0                 10          0  Mid range products   
4                     9.0                 11          1  Mid range products   

     busiest_days  
0  Regularly busy  
1  Regularly busy  
2    Busiest days  
3    Slowest days  
4    Slowest days

## The Column name is changed from 'busiest_day' to 'busiest_days'. The total count of rows remains same. Busiest Days should have a count of (6204182+5660230) = 12916111 for '0' and '1' and it matches. Slowest Days should have a count of (3783802+3840534) = 7624336 for '3' and '4' and it matches. The rest of the count should be (4496490+4213830+4205791) = 12916111 for '6', '2', '5' and it matches.

# 4. Create a column 'busiest_period_of_day' and Label 'order_hour_of_day' As Most orders, Average orders, Fewest orders.

## Frequency Count of 'order_hour_of_day'

In [25]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

## Using For Loops for creating a new list


In [26]:
# Value_counts Between 2761760 and 2000000 = Most orders
# Value_counts Between 2000000 and 400000 = Average orders
# Value_counts Below 300000 = Fewest orders

result_busiest_time_of_day = []

for z in df_ords_prods_merge['order_hour_of_day']:
    if z in [10,11,14,15,13,12,16,9,17 ]:
        result_busiest_time_of_day.append('Most orders')
    elif z in [6,0,1,5,2,4,3]:
        result_busiest_time_of_day.append('Fewest orders')
    else:
        result_busiest_time_of_day.append('Average orders')



In [27]:
# Verifying the new list
result_busiest_time_of_day

['Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Mos

In [28]:
# Comibining with the df_ords_prods_merge
df_ords_prods_merge['busiest_period_of_day'] = result_busiest_time_of_day

In [67]:
# Verifying the dataframe with the added column

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_prior_order  add_to_cart_order  reordered         price_label  \
0                     3.0                  5          0  Mid range products   
1                    20.0                  1          1  Mid range products   
2                     6.0                 20          0  Mid range products   
3                     0.0                 10          0  Mid range products   
4                     9.0                 11          1  Mid range products   

    busiest_days busiest_period_of_day  
0  Regulary busy           Most orders  
1  Regulary busy           Most orders  
2   Busiest days        Average orders  
3   Slowest days           Most orders  
4   Slowest days           Most orders

# 5. Print the frequency of the new column

In [68]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       23205725
Average orders     8312313
Fewest orders       886821
Name: count, dtype: int64

# 6. Export Dataframe

In [70]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_all_dataframes_new_variables.pkl'))